### PR TITLE
Docs: fix and add more examples for new-cap rule (fixes #12874)

### LINKED ISSUE #12874: new cap: option capIsNewExceptionPattern is not behaving as documented
Documentation for `capIsNewExceptionPattern` says the following:
> allows any uppercase-started **function names that match** the specified regex pattern to be called without the new operator.

However, the actual behaviour seems to be that entire [call expression is matched](https://github.com/eslint/eslint/blob/439c83342c364ba3ce5168d54e165b1fe3e35630/lib/rules/new-cap.js#L211) against the pattern, not just the function name.

Either `capIsNewExceptionPattern` should match only function name, or the documentation should reflect the actual behaviour.

I believe matching code with regex is fundamentally wrong and the first option is more sane solution. So, the documentation should stay as is, behaviour corrected and example updated, but that might be a breaking change.

Additionally, [the example](https://eslint.org/docs/rules/new-cap#capisnewexceptionpattern) given as "correct code for this rule" is actually producing warnings (see demo), the reason being the regex configured does not match anything in the code at all.

Also other options of this rule may be poorly documented as well, for example `capIsNewExceptions` also states that it matches names, while actually it can match either name or full expression.

**Tell us about your environment**

Node version: v12.10.0
npm version: v6.10.3
Local ESLint version: v6.8.0 (Currently used)
Global ESLint version: Not found

Also same result on whatever version is currently deployed as demo.

**Please show your full configuration:**

<details>
<summary>Demo links</summary>

Expected no warnings: https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbmV3LWNhcDogW1wiZXJyb3JcIiwgeyBcImNhcElzTmV3RXhjZXB0aW9uUGF0dGVyblwiOiBcIl5CYXIkXCIgfV0qL1xuXG5mb28uQmFyKCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==

Not expected to ignore:
https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbmV3LWNhcDogW1wiZXJyb3JcIiwgeyBcImNhcElzTmV3RXhjZXB0aW9uUGF0dGVyblwiOiBcIl5mb29cXC5CYXIkXCIgfV0qL1xuXG5mb28uQmFyKCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==

Documentation example, wrong on several accounts: https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbmV3LWNhcDogW1wiZXJyb3JcIiwgeyBcImNhcElzTmV3RXhjZXB0aW9uUGF0dGVyblwiOiBcIl5QZXJzb25cXC4uXCIgfV0qL1xuXG52YXIgZnJpZW5kID0gcGVyc29uLkFjcXVhaW50YW5jZSgpO1xudmFyIGJlc3RGcmllbmQgPSBwZXJzb24uRnJpZW5kKCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==
</details>

**What did you do?**

Configuration:
```
eslint new-cap: ["error", { "capIsNewExceptionPattern": "^Bar$" }]
```

Code:
```js
foo.Bar();
```

**What did you expect to happen?**

No warnings as function named `Bar` should be ignored.

**What actually happened? Please include the actual, raw output from ESLint.**

A warning against function `Bar`.

**Are you willing to submit a pull request to fix this bug?**

Only if it's as trivial as `pattern.test(calleeName)`, but then it's probably trivial for anyone else as well.

### PR DESCRIPTION
**Pull request by snitin315**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[X] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Fix incorrect examples and add some more examples.

Fix https://github.com/eslint/eslint/issues/12874
Fix https://github.com/eslint/eslint/pull/13037

#### Is there anything you'd like reviewers to focus on?
No

### CODE REVIEW

File: docs/rules/new-cap.md [hunk 1]
```diff
@@ -135,15 +135,31 @@ function foo(arg) {
...
 var friend = person.Acquaintance();
 var bestFriend = person.Friend();
 ```
 
+Examples of additional **correct** code for this rule with the `{ "capIsNewExceptionPattern": "\.Bar$" }` option:
+
+```js
+/*eslint new-cap: ["error", { "capIsNewExceptionPattern": "\.Bar$" }]*/
+
+foo.Bar();
```
Comment (mdjermanovic) [inline docs/rules/new-cap.md:152]: We should now use double `\\` in all these examples. The new version of `levn` treats content between quotes as JS string literal syntax (https://github.com/eslint/eslint/pull/13140), so `"\.Bar$"` in the directive becomes a string with value `".Bar$"` (`\.` was a useless escape), and then `/.Bar$/` regex, where `.` matches anything, so this configuration directive allows the following:

```js
/*eslint new-cap: ["error", { "capIsNewExceptionPattern": "\.Bar$" }]*/

foo.Bar(); // ok

ABar(); // ok, but that wasn't the intent
```

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbmV3LWNhcDogW1wiZXJyb3JcIiwgeyBcImNhcElzTmV3RXhjZXB0aW9uUGF0dGVyblwiOiBcIlxcLkJhciRcIiB9XSovXG5cbmZvby5CYXIoKTsgLy8gb2tcblxuQUJhcigpOyAvLyBvaywgYnV0IHRoYXQgd2Fzbid0IHRoZSBpbnRlbnQiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

`"\\.Bar$"` translates to `/\.Bar$/` regex, where `\.` matches only dots:

```js
/*eslint new-cap: ["error", { "capIsNewExceptionPattern": "\\.Bar$" }]*/

foo.Bar(); // ok

ABar(); // error
```

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbmV3LWNhcDogW1wiZXJyb3JcIiwgeyBcImNhcElzTmV3RXhjZXB0aW9uUGF0dGVyblwiOiBcIlxcXFwuQmFyJFwiIH1dKi9cblxuZm9vLkJhcigpOyAvLyBva1xuXG5BQmFyKCk7IC8vIGVycm9yIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: docs/rules/new-cap.md (hunk 1 above)
Comment (snitin315) [reply docs/rules/new-cap.md]: Thanks for reviewing. Updated 👍🏻

File: docs/rules/new-cap.md [hunk 2]
```diff
@@ -135,15 +135,31 @@ function foo(arg) {
...
+```
+
+Examples of additional **correct** code for this rule with the `{ "capIsNewExceptionPattern": "^Foo" }` option:
+
+```js
+/*eslint new-cap: ["error", { "capIsNewExceptionPattern": "^Foo" }]*/
+
+var x = Foo(42)
```
Comment (mdjermanovic) [inline docs/rules/new-cap.md:160]:
```suggestion
var x = Foo(42);

var y = Foobar(42);

var z = Foo.Bar(42);
```

to clarify that `"^Foo"` allows more than just `Foo()`

File: docs/rules/new-cap.md [hunk 3]
```diff
@@ -135,15 +135,31 @@ function foo(arg) {
...
 
 ```js
-/*eslint new-cap: ["error", { "capIsNewExceptionPattern": "^Person\.." }]*/
+/*eslint new-cap: ["error", { "capIsNewExceptionPattern": "^person\\.." }]*/
 
 var friend = person.Acquaintance();
 var bestFriend = person.Friend();
 ```
 
+Examples of additional **correct** code for this rule with the `{ "capIsNewExceptionPattern": "\\.Bar$" }` option:
```
Comment (mdjermanovic) [inline docs/rules/new-cap.md:147]: Can you also add a similar example for `"newIsCapExceptionPattern"`?

File: docs/rules/new-cap.md [hunk 4]
```diff
@@ -135,15 +135,31 @@ function foo(arg) {
...
+
+```js
+/*eslint new-cap: ["error", { "capIsNewExceptionPattern": "\\.Bar$" }]*/
```
Comment (mdjermanovic) [inline docs/rules/new-cap.md:150]: Can you also add a similar example for `"newIsCapExceptionPattern"`?

File: docs/rules/new-cap.md (hunk 2 above)
Comment (snitin315) [reply docs/rules/new-cap.md]: Updated 👍🏻

File: docs/rules/new-cap.md (hunk 4 above)
Comment (snitin315) [reply docs/rules/new-cap.md:159]: Added one 👍🏻

File: docs/rules/new-cap.md [hunk 5]
```diff
@@ -112,15 +112,24 @@ var emitter = new events();
...
 
 var friend = new person.acquaintance();
+
 var bestFriend = new person.friend();
 ```
 
+Examples of additional **correct** code for this rule with the `{ "newIsCapExceptionPattern": "\\.Bar$" }` option:
+
+```js
+/*eslint new-cap: ["error", { "newIsCapExceptionPattern": "\\.Bar$" }]*/
```
Comment (mdjermanovic) [inline docs/rules/new-cap.md:128]:
```suggestion
Examples of additional **correct** code for this rule with the `{ "newIsCapExceptionPattern": "\\.bar$" }` option:

```js
/*eslint new-cap: ["error", { "newIsCapExceptionPattern": "\\.bar$" }]*/
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: docs/rules/new-cap.md (hunk 5 above)
Comment (snitin315) [reply docs/rules/new-cap.md:128]: Oh yes!

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**